### PR TITLE
Server systemd notify added similar to Consul. Resolves: #7028

### PR DESCRIPTION
**Pull request by aphorise**
Not sure if there is a more elegant way of reading Systemd configuration so as to only display message conditional to Vault unit file reads:

```
[Service]
Type=notify
```

### CODE REVIEW

File: command/agent.go [hunk 1]
```diff
@@ -15,6 +15,7 @@ import (
 	"sync"
 	"time"
 
+	"github.com/hashicorp/consul/agent/systemd"
```
Comment (ncabatoff) [inline command/agent.go:18]: I don't think we should be depending on an internal Consul package.  Can it be factored out somewhere instead, or if it's very small maybe we can duplicate it?

Comment (aphorise) [reply command/agent.go:18]: It's rather small & I'd agree to making a copy local to this repo retaining some header of where its taken from :-) - let me know what to do.

Comment (aphorise) [reply command/agent.go:18]: @ncabatoff should I localise the `systemd` dependency? - kindly suggest how to proceed with this PR.

Comment (ncabatoff) [reply command/agent.go:18]: Yes please

File: command/agent.go [hunk 2]
```diff
@@ -862,12 +860,11 @@ func (c *AgentCommand) Run(args []string) int {
...
 		c.UI.Error(fmt.Sprintf("Error storing PID: %s", err))
 		return 1
-	} else {
-		if notifyErr := c.NotifierSystemd.Notify(systemd.Ready); notifyErr != nil {
-			c.logger.Debug("server: systemd notify failed", notifyErr)
-		}
 	}
 
+	// Notify systemd that the server is ready (if applicable)
+	c.notifySystemd(systemd.SdNotifyReady)
```
Comment (aphorise) [inline command/agent.go:866]: I tried adding this a bit later after the run condition which may be more valid to NotifyReady as and when it's actually ready - however since the time it may take for that is not as easily determinable then getting an earlier start notification coupled with a followed up by any stop is probably better than just getting NotifyStopping without ever seeing NotifyReady

**Comment by ncabatoff** [comment]
Changes seem fine.  Since this isn't practical to test using Go tests, can you summarize what manual testing you did?

**Comment by aphorise** [comment]
@ncabatoff  thank you for the review.

The manual tests performed were to provision a Vault-agent and to confirm the expected emittance - observing the following tests:

1. Vault agent being started in a valid / proper sense where only `NotifyReady` is expected.
2. Vault agent - later being stopped where only `NotifyStopping` is received.
3. Vault agent being started / with some erroneous condition where it emits ``NotifyReady` attempting to progress as far as possible before crashing / stopping when `NotifyStopping` is then emitted.

**Review by ncabatoff — APPROVED** [review]

**Comment by ncabatoff** [comment]
Thanks!  Could you merge in latest main please?

**Comment by jdoss** [comment]
Any chance to get this rebased and merged?

**Comment by ncabatoff** [comment]
Hi @aphorise, sorry I forgot about this.  Could you merge in main again please, then I'll merge this?

**Comment by aphorise** [comment]
hey @ncabatoff done - hope this is okay. Thanks.

**Comment by aphorise** [comment]
@ncabatoff will this make it into the next release?

File: command/agent.go [hunk 3]
```diff
@@ -794,13 +795,16 @@ func (c *AgentCommand) Run(args []string) int {
...
+				// Notify systemd that the server is shutting down
+				c.notifySystemd(systemd.SdNotifyStopping)
 				// Let the lease cache know this is a shutdown; no need to evict
 				// everything
 				if leaseCache != nil {
 					leaseCache.SetShuttingDown(true)
 				}
 				return nil
 			case <-ctx.Done():
+				c.notifySystemd(systemd.SdNotifyStopping)
```
Comment (ncabatoff) [inline command/agent.go:807]: Shouldn't we also do this in the case above, when we were asked to shutdown?

Comment (aphorise) [reply command/agent.go:807]: hey @ncabatoff I'm not comprehending - does `systemd.SdNotifyStopping` need to be included in any other case?

Comment (aphorise) [reply command/agent.go:807]: If any **one** of the two signals are the only mode of exit (outside Windows) then I think these two later here would suffice and void any repeats - but I'm not entirely sure TBH.

Comment (ncabatoff) [reply command/agent.go:807]: Maybe it's simpler just to add the notifySystemd call to the end of the func, or to a defer.

Comment (aphorise) [reply command/agent.go:807]: Is there any chance that a end or defer function may take longer or not ever reached? - just thinking aloud in case of processor monitors or other cases an agent may be forced quit and never actually ending.

Anyway @ncabatoff any chance you can do a PR if you think a defer may be better.

Comment (ncabatoff) [reply command/agent.go:807]: Defer will always fire when the func returns, and it shouldn't be any slower than doing it just before a return statement.

**Review by ncabatoff — APPROVED** [review]
